spirv_msl.cpp

```diff
@@ -7903,10 +8080,67 @@ void CompilerMSL::emit_glsl_op(uint32_t result_type, uint32_t id, uint32_t eop,
 			emit_trinary_func_op(result_type, id, args[0], args[1], args[2], "precise::clamp");
 		break;
 
-		// TODO:
-		//        GLSLstd450InterpolateAtCentroid (centroid_no_perspective qualifier)
-		//        GLSLstd450InterpolateAtSample (sample_no_perspective qualifier)
-		//        GLSLstd450InterpolateAtOffset
+	case GLSLstd450InterpolateAtCentroid:
+	{
+		// We can't just emit the expression normally, because the qualified name contains a call to the default
+		// interpolate method, or refers to a local variable. We saved the interface index we need; use it to construct
+		// the base for the method call.
+		uint32_t interface_index = get_extended_decoration(args[0], SPIRVCrossDecorationInterfaceMemberIndex);
+		string component;
+		if (has_extended_decoration(args[0], SPIRVCrossDecorationInterpolantComponentExpr))
+		{
+			uint32_t index_expr = get_extended_decoration(args[0], SPIRVCrossDecorationInterpolantComponentExpr);
+			auto *c = maybe_get<SPIRConstant>(index_expr);
+			if (!c || c->specialization)
+				component = join("[", to_expression(index_expr), "]");
```

Comment:
This is questionable. A literal could be interpreted as a SPIRConstant by mistake if that index happened to have a SPIRConstant in it. Perhaps tag the ID with 0x80000000u MSB set to mark literal vs SPIRConstant.

Reply:
This should never be a literal. No `OpAccessChain` instructions in SPIR-V use literals; only the artificial access chains we generate ourselves do, and I don't think we'll wind up calling this on them.

Reply:
Ok, in that case it is misleading to check for `!c` and `maybe_get`.

Reply:
But wait, what if there's an expression, but it's not a `SPIRConstant`? We have other checks like this, e.g. https://github.com/KhronosGroup/SPIRV-Cross/blob/aca9b6879af8af1704e0f54ae31b4118b11d790b/spirv_msl.cpp#L6482

Reply:
I guess ... If it's a dynamic index, then yes.